### PR TITLE
Serve virtctl binaries from the cluster for disconnected environments

### PR DESCRIPTION
**Pull request by erkanerol**
Now, we are redirecting users to github.com download page
when they want to download virtctl binary. However, it doesn't
work in disconnected environment. Therefore, we are starting
serving the binaries from the cluster so that users can download it
in disconnected environments.

Signed-off-by: Erkan Erol <eerol@redhat.com>

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
Serve virtctl binaries from the cluster for disconnected environments
```

### CODE REVIEW

**Review by nunnatsa — APPROVED** [review]
/approve

File: deploy/index-image/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml [hunk 1]
```diff
@@ -2001,6 +2013,40 @@ spec:
                     memory: 48Mi
               priorityClassName: system-node-critical
               serviceAccountName: hyperconverged-cluster-operator
+      - label:
+          app.kubernetes.io/component: deployment
+          app.kubernetes.io/managed-by: olm
+          app.kubernetes.io/part-of: hyperconverged-cluster
+          app.kubernetes.io/version: 1.5.0
+          name: hyperconverged-cluster-webhook
```
Comment (nunnatsa) [inline deploy/index-image/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml:2021]: should be something else. maybe `hyperconverged-cluster-cli-download`?

File: deploy/index-image/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml [hunk 2]
```diff
@@ -2001,6 +2013,40 @@ spec:
...
+        name: downloads
```
Comment (nunnatsa) [inline deploy/index-image/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml:2022]: maybe `hco-cli-download`?

File: deploy/index-image/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml [hunk 3]
```diff
@@ -2001,6 +2013,40 @@ spec:
...
+          strategy:
+            type: RollingUpdate
+          template:
+            metadata:
+              labels:
+                app.kubernetes.io/component: deployment
+                app.kubernetes.io/managed-by: olm
+                app.kubernetes.io/part-of: hyperconverged-cluster
+                app.kubernetes.io/version: 1.5.0
+                name: downloads
```
Comment (nunnatsa) [inline deploy/index-image/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml:2037]: maybe `hco-cli-download`?

File: deploy/olm-catalog/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml [hunk 4]
```diff
@@ -2001,6 +2013,40 @@ spec:
                     memory: 48Mi
               priorityClassName: system-node-critical
               serviceAccountName: hyperconverged-cluster-operator
+      - label:
+          app.kubernetes.io/component: deployment
+          app.kubernetes.io/managed-by: olm
+          app.kubernetes.io/part-of: hyperconverged-cluster
+          app.kubernetes.io/version: 1.5.0
+          name: hyperconverged-cluster-webhook
```
Comment (nunnatsa) [inline deploy/olm-catalog/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml:2021]: should be something else. maybe `hyperconverged-cluster-cli-download`?

File: deploy/olm-catalog/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml [hunk 5]
```diff
@@ -2001,6 +2013,40 @@ spec:
...
+        name: downloads
+        spec:
+          replicas: 1
+          selector:
+            matchLabels:
+              name: downloads
```
Comment (nunnatsa) [inline deploy/olm-catalog/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml:2027]: maybe `hco-cli-download`?

File: deploy/olm-catalog/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml [hunk 6]
```diff
@@ -2001,6 +2013,40 @@ spec:
...
+          strategy:
+            type: RollingUpdate
+          template:
+            metadata:
+              labels:
+                app.kubernetes.io/component: deployment
+                app.kubernetes.io/managed-by: olm
+                app.kubernetes.io/part-of: hyperconverged-cluster
+                app.kubernetes.io/version: 1.5.0
+                name: downloads
```
Comment (nunnatsa) [inline deploy/olm-catalog/community-kubevirt-hyperconverged/1.5.0/manifests/kubevirt-hyperconverged-operator.v1.5.0.clusterserviceversion.yaml:2037]: maybe `hco-cli-download`?

File: deploy/operator.yaml [hunk 7]
```diff
@@ -181,6 +181,42 @@ spec:
 ---
 apiVersion: apps/v1
 kind: Deployment
+metadata:
+  labels:
+    app.kubernetes.io/component: deployment
+    app.kubernetes.io/part-of: hyperconverged-cluster
+    app.kubernetes.io/version: 1.5.0
+    name: downloads
```
Comment (nunnatsa) [inline deploy/operator.yaml:189]: maybe `hco-cli-download`?

File: deploy/operator.yaml [hunk 8]
```diff
@@ -181,6 +181,42 @@ spec:
...
+  name: downloads
```
Comment (nunnatsa) [inline deploy/operator.yaml:190]: maybe `hco-cli-download`?

File: deploy/operator.yaml [hunk 9]
```diff
@@ -181,6 +181,42 @@ spec:
...
+      name: downloads
+  strategy:
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/component: deployment
+        app.kubernetes.io/part-of: hyperconverged-cluster
+        app.kubernetes.io/version: 1.5.0
+        name: downloads
```
Comment (nunnatsa) [inline deploy/operator.yaml:204]: maybe `hco-cli-download`?

File: hack/build-manifests.sh [hunk 10]
```diff
@@ -49,6 +49,7 @@ source "${PROJECT_ROOT}"/deploy/images.env
 
 HCO_OPERATOR_IMAGE=${HCO_OPERATOR_IMAGE:-quay.io/kubevirt/hyperconverged-cluster-operator:${CSV_VERSION}-unstable}
 HCO_WEBHOOK_IMAGE=${HCO_WEBHOOK_IMAGE:-quay.io/kubevirt/hyperconverged-cluster-webhook:${CSV_VERSION}-unstable}
+DOWNLOADS_IMAGE=${DOWNLOADS_IMAGE:-quay.io/kubevirt/virt-artifacts-server:${CSV_VERSION}-unstable}
```
Comment (nunnatsa) [inline hack/build-manifests.sh:52]: `HCO_DOWNLOAD_IMAGE`?

File: hack/build-manifests.sh [hunk 11]
```diff
@@ -49,6 +49,7 @@ source "${PROJECT_ROOT}"/deploy/images.env
...
 DIGEST_LIST="${DIGEST_LIST},${HCO_OPERATOR_IMAGE}|hyperconverged-cluster-operator,${HCO_WEBHOOK_IMAGE}|hyperconverged-cluster-webhook"
```
Comment (nunnatsa) [inline hack/build-manifests.sh:53]: Add the download image to this list

File: pkg/components/components.go [hunk 12]
```diff
@@ -40,13 +40,21 @@ const (
 	hcoWhDeploymentName = "hco-webhook"
 	certVolume          = "apiservice-cert"
 
+	downloadsName = "downloads"
```
Comment (nunnatsa) [inline pkg/components/components.go:43]: maybe `hco-cli-download`?

File: pkg/components/components.go [hunk 13]
```diff
@@ -40,13 +40,21 @@ const (
...
+var deploymentType = metav1.TypeMeta{
+	APIVersion: "apps/v1",
+	Kind:       "Deployment",
+}
+
 type DeploymentOperatorParams struct {
 	Namespace           string
 	Image               string
 	WebhookImage        string
+	DownloadsImage      string
```
Comment (nunnatsa) [inline pkg/components/components.go:57]: `CliDownloadImage`?

File: pkg/components/components.go [hunk 14]
```diff
@@ -800,6 +856,11 @@ func GetInstallStrategyBase(params *DeploymentOperatorParams) *csvv1alpha1.Strat
 				Spec:  GetDeploymentSpecWebhook(params.Namespace, params.WebhookImage, params.ImagePullPolicy, params.HcoKvIoVersion, params.Env),
 				Label: getLabels(hcoNameWebhook, params.HcoKvIoVersion),
 			},
+			{
+				Name:  downloadsName,
+				Spec:  GetDeploymentSpecDownloads(params),
+				Label: getLabels(hcoNameWebhook, params.HcoKvIoVersion),
```
Comment (nunnatsa) [inline pkg/components/components.go:862]: should be `Label: getLabels(downloadsName, params.HcoKvIoVersion),`

File: pkg/controller/operands/cliDownload.go [hunk 15]
```diff
@@ -2,20 +2,29 @@ package operands
...
 	"github.com/kubevirt/hyperconverged-cluster-operator/pkg/controller/common"
 	"github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
 	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
 	consolev1 "github.com/openshift/api/console/v1"
+	routev1 "github.com/openshift/api/route/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	downloadsServiceName = "downloads"
```
Comment (nunnatsa) [inline pkg/controller/operands/cliDownload.go:22]: `cli-downloads`?

**Review by nunnatsa — CHANGES_REQUESTED** [review]
It seems that there is one problem that seems to be a real problem - maybe it's the reason why the installation fails? one of the `name` labels of the new **downloads** deployment, is with the **webhook** name.

In addition, all the rest of the comments are about the name of the new deployment/service/pod etc. : I found `downloads` to be too generic. please consider to rename it to something more specific. And also, the new image name is `virt-artifacts` and it's very different than the `downloads` or even `hco-cli-downloads` and I find it confusing - could we use a similar name that will show the connection between the image and the service/deployment/pod?

**Comment by tiraboschi** [comment]
wait for openshift-ci
/hold

**Comment by erkanerol** [comment]
New PR: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/1487

**Pull request closed without merging**